### ISSUE TITLE
Note: Implementation uses runtime code generation [sticky]

### ISSUE THREAD

**Issue by fabiospampinato** [krausest/js-framework-benchmark#1139]
Some frameworks, like mine, generate code at runtime, usually either via `eval` or via `new Function`.

In theory there could be no problems with that, but from the point of view of a user they must be more confident that something with no runtime code generation at all has no security issues deriving from that compared to something that uses runtime code generation.

There's also a CSP policy for preventing user code from generating code at runtime, as it's a desirable feature for many.

As such I think it would be useful to add an additional flag for marking frameworks that use this technique.

**Comment by WebReflection** [comment]
I was going to propose this flag myself, as people are "*scared*" by events delegation due flags, and not aware their library of choice might not even be usable in the wild due CSP constraints, so that having an *library based on code evaluation* flag would be way more interesting than benchmark specific implementations!

in short, I fully support this new flag and I'd be happy to help out flagging all libraries involved with a cumulative PR.

**Comment by krausest** [comment]
Good idea. I'll add the text to the result table soon and we'll use this issue as the target.
Would be great if you added the issue in package.json where appropriate.
```
  "js-framework-benchmark": {
    ...
    "issues": [
      1139
    ]
  },
```

**Comment by krausest** [comment]
Was accidentally closed.

**Comment by yyx990803** [comment]
Not sure why Vue is flagged in in 018322f697e6906bb068b89ad68acd4610f2e47d - Vue 3 does not use runtime code generation when templates are pre-compiled at build time.

**Comment by WebReflection** [comment]
@yyx990803 this should've been amended after through automated tests, the initial PR simply looked for evaluation, it wasn't super accurate and few got removed (few others got added) later.

**Comment by krausest** [comment]
I switched to an automated check. See https://github.com/krausest/js-framework-benchmark/issues/1148 for details.

**Comment by L3P3** [comment]
What if there is a variant/setting to disable runtime generated code? I think this would be sufficient.
Or should I submit 2 benchmark entries for each variant?
https://github.com/L3P3/lui/issues/50

**Comment by kanashimia** [comment]
Hello, current check for this feature doesn't properly reflect the name or the description of the issue, in particular frameworks that use inline scripts or styles are marked as 1139 which is not as big of a problem as eval or new Function. It probably should be renamed to something like "Implementation doesn't pass strict CSP". Or the check for inline script and styles should be moved into a separate issue, that could be more useful.

**Comment by WebReflection** [comment]
@kanashimia I think the whole point of the flag is to underline indeed that "Implementation doesn't pass strict CSP" ... either it does or it doesn't, that was the point?

**Comment by kanashimia** [comment]
@WebReflection My point is that the title of this issue is misleading and should be changed, what first post here describes sounds more like `default-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'` where "runtime code generation" (unsafe-eval) is forbidden but static inline scripts are allowed. Inline scripts and styles could be safely allowed under strict-dynamic policy with nonces, unlike unsafe-eval. Basically on the results page this note looks like a bigger problem than it is. Although "strict CSP" isn't the most descriptive title too. My guess is that @fabiospampinato wanted specifically a note about the usage of the "unsafe-eval", and it is possible to have that as a different check from the "unsafe-inline" if needed, I will look into implementing that on a weekend.

**Comment by fabiospampinato** [comment]
@kanashimia Yeah I mean if something is safe what's the point of flagging it 🤔 I'm not familiar with these nonces with inline scripts, but I guess it should be accounted for as you say ideally. It might be easier to just update the framework implementation to fetch an external script though.

**Comment by WebReflection** [comment]
@kanashimia 

> Inline scripts and styles could be safely allowed under strict-dynamic policy with nonces, unlike unsafe-eval.

maybe I am outdated in here, but AFAIK with `nonces` you can allow "*unsafe-eval*" too ... right? My thinking is that to keep it simple there should be a flag "*Requires CSP special rules, if possible*" or not. *CSP* is the thing most common users don't even understand/know as a topic, flagging differently on "*yeah, CSP but it's safer!*" or "*CSP no-go*" for those users, would be perceived likely the same, imho, with the counter side-effect this benchmark/project lowers *CSP awareness* as a whole instead (as in: *you can ignore it*)

Let's improve the message, yet let's please try to inform users what they're choosing as a framework, what it needs to work out of rules they won't control once in production, because of foreign hosts they don't control, and so on? thanks!

**Comment by WebReflection** [comment]
late for @L3P3 

> Or should I submit 2 benchmark entries?

this is a benchmark, wouldn't it be wonderful if people could also understand the library/framework constraints and performance under *CSP* rules and outside those?

My gut would say **yes**, any library that has two variants should provide results for both variants, otherwise either one *eval* based result would be misleading, or people won't even try to ship, in 2025, stuff that doesn't require *eval* as a whole (still imho).

**Comment by krausest** [comment]
Ok. Renamed to "Implementation doesn't pass strict CSP" (old name for reference: Implementation uses runtime code generation), which will show up in future result pages too.

**Comment by kanashimia** [comment]
> maybe I am outdated in here, but AFAIK with nonces you can allow "unsafe-eval" too ... right?

That is doable by adding "unsafe-eval" to the CSP but that makes the policy less strict obviously and security aspect much harder to reason about, I believe it was the case for a very brief time that "strict-dynamic" implied "unsafe-eval", but not the case today.
There is an experimental feature called "trusted types" which also allows you to only allow some usually unsafe APIs and only in some cases.

> My thinking is that to keep it simple there should be a flag "Requires CSP special rules, if possible" or not.

Yes.
Now thinking about it deciding on what exact criteria to split this is not straightforward as those `unsafe-*` policies are too forgiving, `nonce` would require modifying framework code, and trusted types policies require JS code, so probably should just keep it simple.

To clarify my thought here also was that runtime code generation specifically changes the possible performance profile of libraries quite a bit, hence a reason for separation, it is very much the case in parsers and validation libraries, but realistically in this benchmark just looking at the results it doesn't look like that matters much.

**Comment by WebReflection** [comment]
@kanashimia 

> There is an experimental feature called "trusted types" which also allows you to only allow some usually unsafe APIs and only in some cases.

until that's usable and "consumable" by this benchmark, and it doesn't require any CSP rule that's a fair thing to state, yet that woudl also pass CSP "*hostile*" automation around this benchmark, until it lands and it's used conveniently, I believe ....

> but realistically in this benchmark just looking at the results it doesn't look like that matters much.

that's the goal of benchmarks ... we're "*splitting hair*" to be one position before, your "*not much matters*" results into a "*chose this, it's faster, look at the bench*" from others PoVs 🤷

All I am saying is that **CSP** or **no need for CSP** should be all it matters in here, the rest requires users investigation around the *CSP* topic, and that's a good thing!

**Comment by wmertens** [comment]
> All I am saying is that CSP or no need for CSP should be all it matters in here, the rest requires users investigation around the CSP topic, and that's a good thing!

To be clear, does this mean that if an implementation with inline scripts also has a nonce embedded, it passes, or not?

**Comment by WebReflection** [comment]
@wmertens can't remember the current automatic check state but basically there are libraries that need CSP precautions and libraries that don't, that's pretty much it, that's the flag, to my understanding.

**Comment by kanashimia** [comment]
@wmertens For context current policy for the check is just this: https://github.com/krausest/js-framework-benchmark/blob/9156eb7f78f3713aa7999f38a69ad4cdb334bd4b/server/src/static/staticRouter.ts#L17
No nonces allowed here.
